compiler/rustc_const_eval/src/interpret/intrinsics.rs

```diff
@@ -643,6 +647,33 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         self.write_bytes_ptr(dst, bytes)
     }
 
+    pub(crate) fn compare_bytes_intrinsic(
+        &mut self,
+        left: &OpTy<'tcx, <M as Machine<'mir, 'tcx>>::Provenance>,
+        right: &OpTy<'tcx, <M as Machine<'mir, 'tcx>>::Provenance>,
+        byte_count: &OpTy<'tcx, <M as Machine<'mir, 'tcx>>::Provenance>,
+    ) -> InterpResult<'tcx, Scalar<M::Provenance>> {
+        let left = self.read_pointer(left)?;
+        let right = self.read_pointer(right)?;
+        let n = Size::from_bytes(self.read_target_usize(byte_count)?);
+        let result = if n == Size::ZERO {
```

Comment:
This shouldn't be needed. Please copy the current memcmp from Miri.

https://github.com/rust-lang/miri/blob/0556996399b8315625df0be4fcf6a7fc0e30a2b0/src/shims/foreign_items.rs#L686-L705

Reply:
Ah, I didn't know that it would already do the right null checks without also requiring "real" alloc ids.  I'll remove the extra checks.

(And yes, that's where I got `read_bytes_ptr_strip_provenance` 🙂)

@rustbot author

Reply:
Done with https://github.com/rust-lang/rust/compare/d29af6737206d64f834a73a185356bf29958ab78..1f4b525312600f65b7c4d93e459bf2a5953ea5d4

Reply:
Looks good. :)

The checks Miri's functions do by default should exactly match what the docs call "valid for reads of size N", that's why you don't need any extra checks. This also ensures we apply the exact same checks everywhere (direct memory accesses, memcpy, memset, memcmp, ...).

Reply:
Nice!  I like the synergy between that and your notes about how to document it.

(More updates coming for the other comments.)